aggie-coding-club/Rev-Registration
# Fix term select after redux-rearrange

## Description

> This PR is based off of #472, which is based on #349

In essence, this fixes all of the issues with term mismatches when switching terms that we discovered in #349, including:

- Course cards' `get_saved_courses` (filling in the course title + card info)
- Course cards' `api/sections` (filling in the sections for sessions)
- Saved schedules
- Availabilities

I would highly suggest going commit-by-commit with this, as I tried to make the commits self-contained in their changes.

Note that the commit `Made ADD_COURSE_CARD reject mismatched terms` (15c0d65) is what fixes the `get_saved_courses` issues from [Ryan's most recent comment](https://github.com/aggie-coding-club/Rev-Registration/pull/349#issuecomment-727628534) on #349.

I must have spent like 15 hours on term select at this point

## Rationale

So there are actually two "bugs" in this, but they're basically the same thing. If you do the steps mentioned in `How to test` for schedules or availabilities, it will correctly ignore the mismatched term's data but will hide the loading indicator b/c the loading indicator is just a boolean state. I think you could fix this by somewhere alongside the `showLoadingIndicator` state (like convert it to a dict or another state variable) store the term, but I mentioned it before and I think we said it was fine. Regardless, here's a gif of it:

![schedules-bug](https://user-images.githubusercontent.com/7295783/101108371-36233200-3589-11eb-8e25-4f5ad05c7dc4.gif)

## How to test

The easiest way to test it is to add `time.sleep(3)` in the respective API view in either `user_sessions/views.py` or `scraper/views.py` for what you're trying to test for.

For instance, if you were testing schedules, you would fill in save 2 different schedules in each term. Next, you'd add a `time.sleep(3)` in `get_saved_schedules` in `users_sessions/views.py`. Then, you would reload the page, have it restore the term and start the loading of the schedules, then change the term and make sure it doesn't restore the initial term's schedules.
